http-driver/src/main/java/org/neo4j/ogm/drivers/http/response/AbstractHttpResponse.java

```diff
@@ -32,81 +35,152 @@
+ * Queries that contain multiple statements with possible a distinct set of columns and statistic, won't work correctly.
+ *
+ * @author Vince Bickers
  * @author Luanne Misquitta
+ * @author Michael J. Simons
  */
-public abstract class AbstractHttpResponse<T> {
+public abstract class AbstractHttpResponse<T> implements AutoCloseable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractHttpResponse.class);
+    private static final JsonFactory JSON_FACTORY = new JsonFactory();
 
-    private final InputStream results;
-    private final JsonParser bufferParser;
     private final ObjectMapper mapper = ObjectMapperFactory.objectMapper();
-    private final TokenBuffer buffer;
+
     private final Class<T> resultClass;
-    private final CloseableHttpResponse httpResponse;
+    private final String[] columns;
+    private final QueryStatistics queryStatistics;
 
-    private String[] columns;
-    private QueryStatistics queryStatistics;
-    private JsonNode responseNode;
-
-    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractHttpResponse.class);
+    private final ResultNodeIterator results;
 
     public AbstractHttpResponse(CloseableHttpResponse httpResponse, Class<T> resultClass) {
+        this(httpResponse, resultClass, true);
+    }
+
+    public AbstractHttpResponse(CloseableHttpResponse httpResponse, Class<T> resultClass, boolean flatMapData) {
 
         this.resultClass = resultClass;
-        try {
-            this.httpResponse = httpResponse;
-            this.results = httpResponse.getEntity().getContent();
-            try (JsonParser parser = ObjectMapperFactory.jsonFactory().createParser(results)) {
-                buffer = new TokenBuffer(parser);
-                //Copy the contents of the response into the token buffer.
-                //This is so that we do not have to serialize the response to textual json while we get to the end of the stream to check for errors
-                parser.nextToken();
-                buffer.copyCurrentStructure(parser);
+
+        try (
+            CloseableHttpResponse httpResponseToClose = httpResponse;
+            InputStream inputStreamOfResponse = httpResponseToClose.getEntity().getContent();
+            TokenBuffer bufferedResponse = createTokenBuffer(inputStreamOfResponse);
+        ) {
+            // First find the errors node without parsing all the other token and throw an exception if necessary
+            JsonParser pointingOnErrors = findNextObject(bufferedResponse.asParserOnFirstToken(), "errors");
+            throwExceptionOnErrorEntry(pointingOnErrors);
+
+            // Find result node and check if it's an array.
+            JsonParser unbufferedResults = findNextObject(bufferedResponse.asParserOnFirstToken(), "results");
+            if (unbufferedResults == null || !JsonToken.START_ARRAY.equals(unbufferedResults.currentToken())) {
+                throw new IOException("Current result object is not an array!");
```

Comment:
In fact this is a true statement for `null` and non-arrays but since we are already here, we could use distinct error messages for each case. The added "complexity" makes it easier for me to also address another fact: the exception throwing a few lines above in the error case happens in a separate method. This would be nice to be the same here so that we have both on the same level of abstraction.